833013 - The port number specified with clientaddr is not used for trap udp socket

commit 521b4e28b4c794a9d6d929858478d13875246ce3
Author: Ivosh <ivosh@users.sourceforge.net>
Date:   Mon Jul 30 10:53:22 2012 -0700

    CHANGES: libnetsnmp: PATCH 3404876: from hardaker: ability to specify local-bound port in addition to address
    ....
        Signed-off-by: Wes Hardaker <hardaker@users.sourceforge.net>

diff -up net-snmp-5.5/include/net-snmp/library/default_store.h.clientaddr-port net-snmp-5.5/include/net-snmp/library/default_store.h
--- net-snmp-5.5/include/net-snmp/library/default_store.h.clientaddr-port	2009-04-01 19:31:07.000000000 +0200
+++ net-snmp-5.5/include/net-snmp/library/default_store.h	2012-09-13 11:16:57.128055631 +0200
@@ -90,6 +90,7 @@ extern          "C" {
 #define NETSNMP_DS_LIB_NO_DISCOVERY        38 /* don't support RFC5343 contextEngineID discovery */
 #define NETSNMP_DS_LIB_TSM_USE_PREFIX      39 /* TSM's simple security name mapping */
 #define NETSNMP_DS_LIB_ALLOW_SELF_SIGNED   40 /* Certs for DTLS */
+#define NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT 42 /* NETSNMP_DS_LIB_CLIENT_ADDR includes address and also port */
 
     /*
      * library integers 
diff -up net-snmp-5.5/perl/default_store/default_store.pm.clientaddr-port net-snmp-5.5/perl/default_store/default_store.pm
--- net-snmp-5.5/perl/default_store/default_store.pm.clientaddr-port	2009-04-23 18:11:12.000000000 +0200
+++ net-snmp-5.5/perl/default_store/default_store.pm	2012-09-13 11:15:51.753300207 +0200
@@ -63,6 +63,7 @@ use vars qw(@ISA %EXPORT_TAGS @EXPORT_OK
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_LOAD
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_SAVE
 				   NETSNMP_DS_LIB_APPEND_LOGFILES
+                                   NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT
 				   NETSNMP_DS_LIB_MIB_WARNINGS
 				   NETSNMP_DS_LIB_SECLEVEL
 				   NETSNMP_DS_LIB_SNMPVERSION
@@ -161,6 +162,7 @@ use vars qw(@ISA %EXPORT_TAGS @EXPORT_OK
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_LOAD
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_SAVE
 				   NETSNMP_DS_LIB_APPEND_LOGFILES
+                                   NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT
 				   NETSNMP_DS_LIB_MIB_WARNINGS
 				   NETSNMP_DS_LIB_SECLEVEL
 				   NETSNMP_DS_LIB_SNMPVERSION
@@ -299,6 +301,7 @@ None by default.
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_LOAD
 				   NETSNMP_DS_LIB_DISABLE_PERSISTENT_SAVE
 				   NETSNMP_DS_LIB_APPEND_LOGFILES
+                                   NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT
 				   NETSNMP_DS_LIB_MIB_WARNINGS
 				   NETSNMP_DS_LIB_SECLEVEL
 				   NETSNMP_DS_LIB_SNMPVERSION
diff -up net-snmp-5.5/perl/default_store/default_store.xs.clientaddr-port net-snmp-5.5/perl/default_store/default_store.xs
--- net-snmp-5.5/perl/default_store/default_store.xs.clientaddr-port	2006-10-18 01:53:13.000000000 +0200
+++ net-snmp-5.5/perl/default_store/default_store.xs	2012-09-13 11:15:51.754300204 +0200
@@ -1233,6 +1233,17 @@ __END__
 #endif
     }
     break;
+  case 36:
+    if (memEQ(name, "NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT", 36)) {
+#ifdef NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT
+      *iv_return = NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT;
+      return PERL_constant_ISIV;
+#else
+      return PERL_constant_NOTDEF;
+#endif
+    }
+    break;
+
   case 38:
     return constant_38 (aTHX_ name, iv_return);
     break;
@@ -1240,8 +1251,6 @@ __END__
   return PERL_constant_NOTFOUND;
 }
 
-
-
 /* autogenerated by "gen" from const-xs.inc */
 
 MODULE = NetSNMP::default_store         PACKAGE = NetSNMP::default_store
diff -up net-snmp-5.5/perl/default_store/test.pl.clientaddr-port net-snmp-5.5/perl/default_store/test.pl
--- net-snmp-5.5/perl/default_store/test.pl.clientaddr-port	2006-10-21 14:09:09.000000000 +0200
+++ net-snmp-5.5/perl/default_store/test.pl	2012-09-13 11:15:51.755300201 +0200
@@ -52,6 +52,7 @@ BEGIN { $| = 1;
                   "NETSNMP_DS_LIB_DISABLE_PERSISTENT_LOAD" => 35,
                   "NETSNMP_DS_LIB_DISABLE_PERSISTENT_SAVE" => 36,
                   "NETSNMP_DS_LIB_APPEND_LOGFILES"         => 37,
+                  "NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT"   => 42,
                   "NETSNMP_DS_LIB_MIB_WARNINGS"            => 0,
                   "NETSNMP_DS_LIB_SECLEVEL"                => 1,
                   "NETSNMP_DS_LIB_SNMPVERSION"             => 2,
diff -up net-snmp-5.5/snmplib/snmp_api.c.clientaddr-port net-snmp-5.5/snmplib/snmp_api.c
--- net-snmp-5.5/snmplib/snmp_api.c.clientaddr-port	2012-09-13 11:15:51.731300291 +0200
+++ net-snmp-5.5/snmplib/snmp_api.c	2012-09-13 11:15:51.760300181 +0200
@@ -758,6 +758,8 @@ register_default_handlers(void)
 	              NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_16BIT_IDS);
-    netsnmp_ds_register_config(ASN_OCTET_STR, "snmp", "clientaddr",
+    netsnmp_ds_register_premib(ASN_OCTET_STR, "snmp", "clientaddr",
                       NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_CLIENT_ADDR);
+    netsnmp_ds_register_premib(ASN_BOOLEAN, "snmp", "clientaddrUsesPort",
+                      NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT);
     netsnmp_ds_register_config(ASN_INTEGER, "snmp", "serverSendBuf",
 		      NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_SERVERSENDBUF);
     netsnmp_ds_register_config(ASN_INTEGER, "snmp", "serverRecvBuf",
diff -up net-snmp-5.5/snmplib/snmpUDPDomain.c.clientaddr-port net-snmp-5.5/snmplib/snmpUDPDomain.c
--- net-snmp-5.5/snmplib/snmpUDPDomain.c.clientaddr-port	2012-09-13 11:15:51.000000000 +0200
+++ net-snmp-5.5/snmplib/snmpUDPDomain.c	2012-09-13 11:19:28.870493750 +0200
@@ -703,7 +703,28 @@ netsnmp_udp_transport(struct sockaddr_in
                                               NETSNMP_DS_LIB_CLIENT_ADDR);
         if (client_socket) {
             struct sockaddr_in client_addr;
-            netsnmp_sockaddr_in2(&client_addr, client_socket, NULL);
+            char *client_address = client_socket;
+            int uses_port = netsnmp_ds_get_boolean(NETSNMP_DS_LIBRARY_ID,
+                                                   NETSNMP_DS_LIB_CLIENT_ADDR_USES_PORT);
+            if ((uses_port == 1) && (strchr(client_socket, ':') == NULL)) {
+                client_address = malloc(strlen(client_socket) + 3);
+                if (client_address == NULL) {
+                    netsnmp_udp_close(t);
+                    netsnmp_transport_free(t);
+                    return NULL;
+                }                                      /* if NETSNMP_DS_LIB_CLIENT_ADDR */
+                strcpy(client_address, client_socket); /* expects a port but there is none */
+                strcat(client_address, ":0");          /* specified then provide ephemeral one */
+            }
+
+            netsnmp_sockaddr_in2(&client_addr, client_address, NULL);
+            if (uses_port == 0) {
+                client_addr.sin_port = 0;
+            }
+            if (client_address != client_socket) {
+                free(client_address);
+            }
+
             addr_pair.local_addr = client_addr.sin_addr;
             rc = bind(t->sock, (struct sockaddr *)&client_addr,
                   sizeof(struct sockaddr));
commit 9e00fff692081e36c9d883fab7b6bd8881c670fc
Author: Jan Safranek <jsafranek@users.sourceforge.net>
Date:   Tue Aug 7 12:27:18 2012 +0200

    Document new clientaddrUsesPort option.

diff --git a/man/snmp.conf.5.def b/man/snmp.conf.5.def
index 9c7c55b..904635b 100644
--- a/man/snmp.conf.5.def
+++ b/man/snmp.conf.5.def
@@ -127,6 +127,10 @@ This value is also used by \fBsnmpd\fR when generating notifications.
 .\"  But not responses to an incoming request?
 .\"  What about snmptrapd?
 .\"
+.IP "clientaddrUsesPort no"
+specifies, if clientaddr option contains a port number. Set this option
+to "yes", if clientaddr contains a port number and this port should
+be used for sending outgoing SNMP requests.
 .IP "clientRecvBuf INTEGER"
 specifies the desired size of the buffer to be used when receiving
 responses to SNMP requests.
